App using programmatic APIs fails to typecheck with `exactOptionalPropertyTypes: true` and `@types/node` v24

When `skipLibCheck` is disabled (the default), a minimal app that is importing from the API entry points fails to compile when:
- `exactOptionalPropertyTypes` is enabled
- `@types/node` v24 or lower is installed

---

Reported errors:

```shell
node_modules/typescript/dist/api/node/node.d.ts:13:24 - error TS2749: 'TextDecoder' refers to a value, but is being used as a type here. Did you mean 'typeof TextDecoder'?

13     readonly _decoder: TextDecoder;
                          ~~~~~~~~~~~

node_modules/typescript/dist/api/node/node.d.ts:14:44 - error TS2749: 'TextDecoder' refers to a value, but is being used as a type here. Did you mean 'typeof TextDecoder'?

14     constructor(data: Uint8Array, decoder: TextDecoder);
                                              ~~~~~~~~~~~

node_modules/typescript/dist/api/node/node.generated.d.ts:3:84 - error TS2344: Type 'RemoteNode' does not satisfy the constraint 'Node'.
  Types of property 'jsDoc' are incompatible.
    Type 'readonly Node[] | undefined' is not assignable to type 'readonly Node[]'.
      Type 'undefined' is not assignable to type 'readonly Node[]'.

3 export declare class RemoteNodeList extends Array<RemoteNode> implements NodeArray<RemoteNode> {
                                                                                     ~~~~~~~~~~

node_modules/typescript/dist/api/node/node.generated.d.ts:38:22 - error TS2420: Class 'RemoteNode' incorrectly implements interface 'Node'.
  Types of property 'jsDoc' are incompatible.
    Type 'readonly Node[] | undefined' is not assignable to type 'readonly Node[]'.
      Type 'undefined' is not assignable to type 'readonly Node[]'.

38 export declare class RemoteNode extends RemoteNodeBase implements Node {
                        ~~~~~~~~~~

node_modules/typescript/dist/api/node/node.infrastructure.d.ts:15:24 - error TS2749: 'TextDecoder' refers to a value, but is being used as a type here. Did you mean 'typeof TextDecoder'?

15     readonly _decoder: TextDecoder;
                          ~~~~~~~~~~~
```

---

To reproduce in this repo:
- include `"exactOptionalPropertyTypes": true` in `_packages/native-preview/tsconfig.json`
- install `@types/node` v24 or lower in `_packages/native-preview/package.json`

--- a/_packages/native-preview/src/api/async/api.ts
+++ b/_packages/native-preview/src/api/async/api.ts
@@ -1447,9 +1447,9 @@ export class Checker {
 }
 
 export interface PrintNodeOptions {
-    preserveSourceNewlines?: boolean;
-    neverAsciiEscape?: boolean;
-    terminateUnterminatedLiterals?: boolean;
+    preserveSourceNewlines?: boolean | undefined;
+    neverAsciiEscape?: boolean | undefined;
+    terminateUnterminatedLiterals?: boolean | undefined;
 }
 
 export class Emitter {
@@ -1499,7 +1499,7 @@ export interface ReferencedSymbolEntry {
     /** The node handle for the symbol's definition. */
     definition: NodeHandle;
     /** The resolved symbol for the definition, if available. */
-    symbol?: Symbol;
+    symbol?: Symbol | undefined;
     /** The node handles for each reference to the symbol. */
     references: NodeHandle[];
 }
@@ -1509,7 +1509,7 @@ export interface SignatureUsage {
     /** The node handle for the name reference. */
     name: NodeHandle;
     /** The node handle for the call expression, if the reference is invoked. */
-    call?: NodeHandle;
+    call?: NodeHandle | undefined;
 }
 
 export class Symbol {

--- a/_packages/native-preview/src/api/async/types.ts
+++ b/_packages/native-preview/src/api/async/types.ts
@@ -177,7 +177,7 @@ export interface IntersectionType extends UnionOrIntersectionType {
 /** Type parameters (TypeFlags.TypeParameter) */
 export interface TypeParameter extends Type {
     /** True if this is the synthetic `this` type of an interface, class, or tuple */
-    readonly isThisType?: boolean;
+    readonly isThisType?: boolean | undefined;
 }
 
 /** Index types — keyof T (TypeFlags.Index) */
@@ -282,7 +282,7 @@ export interface IndexInfo {
     /** Whether the index signature is readonly */
     readonly isReadonly: boolean;
     /** The index signature declaration, if any */
-    readonly declaration?: NodeHandle;
+    readonly declaration?: NodeHandle | undefined;
 }
 
 /**
@@ -292,32 +292,32 @@ export interface JSDocTagInfo {
     /** The tag name, without the leading `@` — e.g. `"param"`. */
     readonly name: string;
     /** The rendered tag text, if any — e.g. `"a the first number"` for `@param a the first number`. */
-    readonly text?: string;
+    readonly text?: string | undefined;
 }
 
 export interface CompletionEntryLabelDetails {
-    detail?: string;
-    description?: string;
+    detail?: string | undefined;
+    description?: string | undefined;
 }
 
 /** Options for {@link Checker.getCompletionsAtPosition}. */
 export interface CompletionOptions {
-    triggerCharacter?: string;
+    triggerCharacter?: string | undefined;
     /** Include a `symbol` property on each completion entry. Only populated for symbol-based completions (not keywords or literals). */
-    includeSymbol?: boolean;
+    includeSymbol?: boolean | undefined;
 }
 
 /** A single completion item returned by {@link Checker.getCompletionsAtPosition}. */
 export interface CompletionEntry {
     readonly name: string;
-    readonly kind?: CompletionItemKind;
-    readonly sortText?: string;
-    readonly insertText?: string;
-    readonly filterText?: string;
-    readonly detail?: string;
-    readonly labelDetails?: CompletionEntryLabelDetails;
+    readonly kind?: CompletionItemKind | undefined;
+    readonly sortText?: string | undefined;
+    readonly insertText?: string | undefined;
+    readonly filterText?: string | undefined;
+    readonly detail?: string | undefined;
+    readonly labelDetails?: CompletionEntryLabelDetails | undefined;
     /** The symbol associated with this completion entry. Only set when `includeSymbol: true` is passed and a symbol is available. */
-    readonly symbol?: Symbol;
+    readonly symbol?: Symbol | undefined;
 }
 
 /** The result of {@link Checker.getCompletionsAtPosition}. */
@@ -331,7 +331,7 @@ export interface CompletionInfo {
  */
 export interface Diagnostic {
     /** File name of the source file this diagnostic belongs to, if any */
-    readonly fileName?: string;
+    readonly fileName?: string | undefined;
     /** Start position of the diagnostic */
     readonly pos: number;
     /** End position of the diagnostic */
@@ -343,11 +343,11 @@ export interface Diagnostic {
     /** Localized diagnostic message text */
     readonly text: string;
     /** Whether this diagnostic highlights unnecessary code */
-    readonly reportsUnnecessary?: boolean;
+    readonly reportsUnnecessary?: boolean | undefined;
     /** Whether this diagnostic highlights deprecated code */
-    readonly reportsDeprecated?: boolean;
+    readonly reportsDeprecated?: boolean | undefined;
     /** Chained diagnostic messages */
-    readonly messageChain?: readonly Diagnostic[];
+    readonly messageChain?: readonly Diagnostic[] | undefined;
     /** Related diagnostic information */
-    readonly relatedInformation?: readonly Diagnostic[];
+    readonly relatedInformation?: readonly Diagnostic[] | undefined;
 }

--- a/_packages/native-preview/src/api/node/node.infrastructure.ts
+++ b/_packages/native-preview/src/api/node/node.infrastructure.ts
@@ -2,7 +2,6 @@ import {
     type FileReference,
     ModifierFlags,
     type Node,
-    type NodeArray,
     SyntaxKind,
 } from "../../ast/index.ts";
 import {
@@ -38,6 +37,11 @@ export const NODE_EXTENDED_DATA_MASK = 0x00_ff_ff_ff;
 // SourceFileInfo — the interface RemoteNode/RemoteNodeList need from the
 // source file, avoiding a direct dependency on RemoteSourceFile.
 // ═══════════════════════════════════════════════════════════════════════════
+
+// The global type is not available in earlier @types/node versions
+export interface TextDecoder {
+    decode(input?: ArrayBufferView | ArrayBufferLike): string;
+}
 
 export interface SourceFileInfo {
     readonly _offsetNodes: number;

--- a/_packages/native-preview/src/api/node/node.ts
+++ b/_packages/native-preview/src/api/node/node.ts
@@ -14,6 +14,7 @@ import {
 import {
     NODE_EXTENDED_DATA_MASK,
     type SourceFileInfo,
+    type TextDecoder,
 } from "./node.infrastructure.ts";
 import {
     HEADER_OFFSET_EXTENDED_DATA,

--- a/_packages/native-preview/src/api/sync/api.ts
+++ b/_packages/native-preview/src/api/sync/api.ts
@@ -117,7 +117,7 @@ import type {
 } from "./types.ts";
 
 export { CompletionItemKind, DiagnosticCategory, ElementFlags, ModifierFlags, ModuleKind, NodeBuilderFlags, ObjectFlags, SignatureFlags, SignatureKind, SymbolFlags, TypeFlags, TypePredicateKind };
-export type { __String, APIOptions, ClientSocketOptions, ClientSpawnOptions, DocumentIdentifier, DocumentPosition, LSPConnectionOptions, SourceFileMetadata };
+export type { APIOptions, ClientSocketOptions, ClientSpawnOptions, DocumentIdentifier, DocumentPosition, LSPConnectionOptions, SourceFileMetadata };
 export type { AssertsIdentifierTypePredicate, AssertsThisTypePredicate, BigIntLiteralType, BooleanLiteralType, CompletionEntry, CompletionInfo, CompletionOptions, ConditionalType, Diagnostic, FreshableType, IdentifierTypePredicate, IndexedAccessType, IndexInfo, IndexType, InterfaceType, IntersectionType, IntrinsicType, JSDocTagInfo, LiteralType, NumberLiteralType, ObjectType, StringLiteralType, StringMappingType, SubstitutionType, TemplateLiteralType, ThisTypePredicate, TupleType, Type, TypeParameter, TypePredicate, TypePredicateBase, TypeReference, UnionOrIntersectionType, UnionType };
 export { documentURIToFileName, fileNameToDocumentURI } from "../path.ts";
 

--- a/_packages/native-preview/src/api/sync/api.ts
+++ b/_packages/native-preview/src/api/sync/api.ts
@@ -117,7 +117,7 @@ import type {
 } from "./types.ts";
 
 export { CompletionItemKind, DiagnosticCategory, ElementFlags, ModifierFlags, ModuleKind, NodeBuilderFlags, ObjectFlags, SignatureFlags, SignatureKind, SymbolFlags, TypeFlags, TypePredicateKind };
-export type { APIOptions, ClientSocketOptions, ClientSpawnOptions, DocumentIdentifier, DocumentPosition, LSPConnectionOptions, SourceFileMetadata };
+export type { __String, APIOptions, ClientSocketOptions, ClientSpawnOptions, DocumentIdentifier, DocumentPosition, LSPConnectionOptions, SourceFileMetadata };
 export type { AssertsIdentifierTypePredicate, AssertsThisTypePredicate, BigIntLiteralType, BooleanLiteralType, CompletionEntry, CompletionInfo, CompletionOptions, ConditionalType, Diagnostic, FreshableType, IdentifierTypePredicate, IndexedAccessType, IndexInfo, IndexType, InterfaceType, IntersectionType, IntrinsicType, JSDocTagInfo, LiteralType, NumberLiteralType, ObjectType, StringLiteralType, StringMappingType, SubstitutionType, TemplateLiteralType, ThisTypePredicate, TupleType, Type, TypeParameter, TypePredicate, TypePredicateBase, TypeReference, UnionOrIntersectionType, UnionType };
 export { documentURIToFileName, fileNameToDocumentURI } from "../path.ts";
 
@@ -1455,9 +1455,9 @@ export class Checker {
 }
 
 export interface PrintNodeOptions {
-    preserveSourceNewlines?: boolean;
-    neverAsciiEscape?: boolean;
-    terminateUnterminatedLiterals?: boolean;
+    preserveSourceNewlines?: boolean | undefined;
+    neverAsciiEscape?: boolean | undefined;
+    terminateUnterminatedLiterals?: boolean | undefined;
 }
 
 export class Emitter {
@@ -1507,7 +1507,7 @@ export interface ReferencedSymbolEntry {
     /** The node handle for the symbol's definition. */
     definition: NodeHandle;
     /** The resolved symbol for the definition, if available. */
-    symbol?: Symbol;
+    symbol?: Symbol | undefined;
     /** The node handles for each reference to the symbol. */
     references: NodeHandle[];
 }
@@ -1517,7 +1517,7 @@ export interface SignatureUsage {
     /** The node handle for the name reference. */
     name: NodeHandle;
     /** The node handle for the call expression, if the reference is invoked. */
-    call?: NodeHandle;
+    call?: NodeHandle | undefined;
 }
 
 export class Symbol {

--- a/_packages/native-preview/src/api/sync/types.ts
+++ b/_packages/native-preview/src/api/sync/types.ts
@@ -185,7 +185,7 @@ export interface IntersectionType extends UnionOrIntersectionType {
 /** Type parameters (TypeFlags.TypeParameter) */
 export interface TypeParameter extends Type {
     /** True if this is the synthetic `this` type of an interface, class, or tuple */
-    readonly isThisType?: boolean;
+    readonly isThisType?: boolean | undefined;
 }
 
 /** Index types — keyof T (TypeFlags.Index) */
@@ -290,7 +290,7 @@ export interface IndexInfo {
     /** Whether the index signature is readonly */
     readonly isReadonly: boolean;
     /** The index signature declaration, if any */
-    readonly declaration?: NodeHandle;
+    readonly declaration?: NodeHandle | undefined;
 }
 
 /**
@@ -300,32 +300,32 @@ export interface JSDocTagInfo {
     /** The tag name, without the leading `@` — e.g. `"param"`. */
     readonly name: string;
     /** The rendered tag text, if any — e.g. `"a the first number"` for `@param a the first number`. */
-    readonly text?: string;
+    readonly text?: string | undefined;
 }
 
 export interface CompletionEntryLabelDetails {
-    detail?: string;
-    description?: string;
+    detail?: string | undefined;
+    description?: string | undefined;
 }
 
 /** Options for {@link Checker.getCompletionsAtPosition}. */
 export interface CompletionOptions {
-    triggerCharacter?: string;
+    triggerCharacter?: string | undefined;
     /** Include a `symbol` property on each completion entry. Only populated for symbol-based completions (not keywords or literals). */
-    includeSymbol?: boolean;
+    includeSymbol?: boolean | undefined;
 }
 
 /** A single completion item returned by {@link Checker.getCompletionsAtPosition}. */
 export interface CompletionEntry {
     readonly name: string;
-    readonly kind?: CompletionItemKind;
-    readonly sortText?: string;
-    readonly insertText?: string;
-    readonly filterText?: string;
-    readonly detail?: string;
-    readonly labelDetails?: CompletionEntryLabelDetails;
+    readonly kind?: CompletionItemKind | undefined;
+    readonly sortText?: string | undefined;
+    readonly insertText?: string | undefined;
+    readonly filterText?: string | undefined;
+    readonly detail?: string | undefined;
+    readonly labelDetails?: CompletionEntryLabelDetails | undefined;
     /** The symbol associated with this completion entry. Only set when `includeSymbol: true` is passed and a symbol is available. */
-    readonly symbol?: Symbol;
+    readonly symbol?: Symbol | undefined;
 }
 
 /** The result of {@link Checker.getCompletionsAtPosition}. */
@@ -339,7 +339,7 @@ export interface CompletionInfo {
  */
 export interface Diagnostic {
     /** File name of the source file this diagnostic belongs to, if any */
-    readonly fileName?: string;
+    readonly fileName?: string | undefined;
     /** Start position of the diagnostic */
     readonly pos: number;
     /** End position of the diagnostic */
@@ -351,11 +351,11 @@ export interface Diagnostic {
     /** Localized diagnostic message text */
     readonly text: string;
     /** Whether this diagnostic highlights unnecessary code */
-    readonly reportsUnnecessary?: boolean;
+    readonly reportsUnnecessary?: boolean | undefined;
     /** Whether this diagnostic highlights deprecated code */
-    readonly reportsDeprecated?: boolean;
+    readonly reportsDeprecated?: boolean | undefined;
     /** Chained diagnostic messages */
-    readonly messageChain?: readonly Diagnostic[];
+    readonly messageChain?: readonly Diagnostic[] | undefined;
     /** Related diagnostic information */
-    readonly relatedInformation?: readonly Diagnostic[];
+    readonly relatedInformation?: readonly Diagnostic[] | undefined;
 }

--- a/_packages/native-preview/src/ast/ast.ts
+++ b/_packages/native-preview/src/ast/ast.ts
@@ -55,7 +55,7 @@ export interface Node extends ReadonlyTextRange {
     readonly kind: SyntaxKind;
     readonly flags: NodeFlags;
     readonly parent: Node;
-    readonly jsDoc?: readonly Node[];
+    readonly jsDoc?: readonly Node[] | undefined;
     forEachChild<T>(visitor: (node: Node) => T, visitArray?: (nodes: NodeArray<Node>) => T): T | undefined;
     getSourceFile(): SourceFile;
     getStart(sourceFile?: SourceFile, includeJsDocComment?: boolean): number;

--- a/_packages/native-preview/test/tsconfig.json
+++ b/_packages/native-preview/test/tsconfig.json
@@ -1,5 +1,9 @@
 {
     "extends": "../tsconfig.dev.json",
+    "compilerOptions": {
+        // tinybench has errors under exactOptionalPropertyTypes
+        "skipLibCheck": true
+    },
     "references": [
         { "path": ".." }
     ]

--- a/_packages/native-preview/tsconfig.base.json
+++ b/_packages/native-preview/tsconfig.base.json
@@ -3,6 +3,7 @@
         "target": "es2022",
         "lib": ["es2022"],
         "strict": true,
+        "exactOptionalPropertyTypes": true,
         "composite": true,
         "forceConsistentCasingInFileNames": true,
         "rewriteRelativeImportExtensions": true,

